### PR TITLE
Makes text selection match the native behavior

### LINKED ISSUE #79166: SelectableText double click on the space before a word to select the previous word throws if there's no word
### to reproduce

1. create a SelectableText with string which ends on word and begins with two or more spaces, e.g. "  word"
2. double click the left half of the space that goes right before the word - it will try to select the previous word but will fail because there's isn't any

in my code sample you should click the left half of the green box

image

<details>
<summary>logs</summary>

```
══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
The following _CastError was thrown while handling a gesture:
Null check operator used on a null value
When the exception was thrown, this was the stack:
#0      RenderEditable._getWordAtOffset (package:flutter/src/rendering/editable.dart:3061:39)
#1      RenderEditable.selectWordsInRange (package:flutter/src/rendering/editable.dart:3003:37)
#2      RenderEditable.selectWord (package:flutter/src/rendering/editable.dart:2989:5)
#3      TextSelectionGestureDetectorBuilder.onDoubleTapDown (package:flutter/src/widgets/text_selection.dart:1171:22)
#4      _TextSelectionGestureDetectorState._handleTapDown (package:flutter/src/widgets/text_selection.dart:1396:31)
#5      TapGestureRecognizer.handleTapDown.<anonymous closure> (package:flutter/src/gestures/tap.dart:580:61)
#6      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
#7      TapGestureRecognizer.handleTapDown (package:flutter/src/gestures/tap.dart:580:11)
#8      BaseTapGestureRecognizer._checkDown (package:flutter/src/gestures/tap.dart:287:5)
#9      BaseTapGestureRecognizer.didExceedDeadline (package:flutter/src/gestures/tap.dart:258:5)
#10     PrimaryPointerGestureRecognizer.didExceedDeadlineWithEvent (package:flutter/src/gestures/recognizer.dart:501:5)
#11     PrimaryPointerGestureRecognizer.addAllowedPointer.<anonymous closure> (package:flutter/src/gestures/recognizer.dart:454:41)
(elided 11 frames from class _RawReceivePortImpl, class _Timer, dart:async, and dart:async-patch)
Handler: "onTapDown"
Recognizer:
  _TransparentTapGestureRecognizer#51bba
════════════════════════════════════════════════════════════════════════════════════════════════════
```
</details>

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    const title = 'code sample';
    return MaterialApp(
      title: title,
      home: Scaffold(
        backgroundColor: Colors.black,
        body: Home(),
      ),
    );
  }
}

class Home extends StatelessWidget {
  const Home({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: SingleChildScrollView(
        child: Theme(
          data: ThemeData(
            textSelectionTheme: TextSelectionThemeData(selectionColor: Colors.red),
          ),
          child: SelectableText.rich(
            TextSpan(
              children: [
                TextSpan(text: ' ',
                  style: TextStyle(
                    fontSize: 110.0,
                  ),
                ),
                TextSpan(text: ' ',
                  style: TextStyle(
                    backgroundColor: Colors.green,
                    fontSize: 110.0,
                  ),
                ),
                TextSpan(text: 'word',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 110.0,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 2.1.0-13.0.pre.288, on Microsoft Windows [Version 10.0.19041.867], locale ru-RU)
    • Flutter version 2.1.0-13.0.pre.288 at c:\dev\src\flutter
    • Framework revision e384ca7979 (80 minutes ago), 2021-03-26 17:34:03 +0100
    • Engine revision b5e15d055d
    • Dart version 2.13.0 (build 2.13.0-162.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)    
    • Android SDK at C:\Users\danya\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)   
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.7)       
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30621.155
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)   

[√] VS Code (version 1.54.3)
    • VS Code at C:\Users\danya\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.20.0

[√] Connected device (4 available)
    • Redmi Note 5 (mobile) • 40c9f14 • android-arm64  • Android 9 (API 28)
    • Windows (desktop)     • windows • windows-x64    • Microsoft Windows [Version 10.0.19041.867]
    • Chrome (web)          • chrome  • web-javascript • Google Chrome 89.0.4389.90
    • Edge (web)            • edge    • web-javascript • Microsoft Edge 89.0.774.57

• No issues found!
```
</details>

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/79166

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -3043,22 +3043,43 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
       return TextSelection(baseOffset: 0, extentOffset: _plainText.length);
-    // If the word is a space, on iOS try to select the previous word instead.
-    // On Android try to select the previous word instead only if the text is read only.
+    // On iOS, select the previous word if there is a previous word, or select
+    // the the whitespace to the next word if there is a next word. Select
+    // nothing if there is neither a previous word nor a next word.
+    //
+    // If the platform is Android and the text is read only, try to select the
+    // previous word if there is one; otherwise, select the whitespace at the
+    // position.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:3052]:
```diff
- the whitespace
+ the single whitespace
```

Just to clarify that it won't select a bunch of whitespace, just one space character (right?).

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -3043,22 +3043,43 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     // If text is obscured, the entire sentence should be treated as one word.
     if (obscureText) {
       return TextSelection(baseOffset: 0, extentOffset: _plainText.length);
-    // If the word is a space, on iOS try to select the previous word instead.
-    // On Android try to select the previous word instead only if the text is read only.
+    // On iOS, select the previous word if there is a previous word, or select
+    // the the whitespace to the next word if there is a next word. Select
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:3047]: Double "the the" here.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -3043,22 +3043,43 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
     } else if (_isWhitespace(_plainText.codeUnitAt(position.offset))
         && position.offset > 0) {
       assert(defaultTargetPlatform != null);
       final TextRange? previousWord = _getPreviousWord(word.start);
       switch (defaultTargetPlatform) {
         case TargetPlatform.iOS:
+          if (previousWord == null) {
+            final TextRange? nextWord = _getNextWord(word.start);
+            if (nextWord == null) {
+              return TextSelection.collapsed(offset:position.offset);
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:3062]: Missing a space here: `offset: position.offset`.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:3047]:
```diff
- the the whitespace to the next word if there is a next word.
+ to the end of the next word if there is a next word.
```

File: packages/flutter/lib/src/rendering/editable.dart [hunk 4]
```diff
@@ -3043,22 +3043,43 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
-            baseOffset: previousWord!.start,
+            baseOffset: previousWord.start,
             extentOffset: position.offset,
           );
         case TargetPlatform.android:
           if (readOnly) {
+            if (previousWord == null) {
+              return TextSelection(
+                baseOffset: position.offset,
+                extentOffset: position.offset + 1,
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:3078]: Could it ever happen than position.offset + 1 is beyond the end of the text?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just some docs improvements below, and one potential out of bounds problem that I'm not sure about.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 4 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/editable.dart:3078]: It is not possible, the case where user select the end of the text is handled in line 3041. so it is guaranteed the position.offset < text.length

**Pull request merged**